lib/backend/firestore/firestorebk.go

```diff
@@ -520,33 +531,26 @@ func (b *Backend) CompareAndSwap(ctx context.Context, expected backend.Item, rep
 		return nil, trace.BadParameter("expected and replaceWith keys should match")
 	}
 
-	ref := b.svc.Collection(b.CollectionName).Doc(b.keyToDocumentID(expected.Key))
-	err := b.svc.RunTransaction(ctx, func(ctx context.Context, tx *firestore.Transaction) error {
-		expectedDocSnap, err := tx.Get(ref)
-		if err != nil {
-			return trace.CompareFailed("error or object not found, error: %v", ConvertGRPCError(err))
-		}
-
-		existingRecord, err := newRecordFromDoc(expectedDocSnap)
-		if err != nil {
-			return trace.Wrap(err)
-		}
-
-		if !bytes.Equal(existingRecord.Value, expected.Value) {
-			return trace.CompareFailed("expected item value %v does not match actual item value %v", string(expected.Value), existingRecord.Value)
-		}
-
-		if err := tx.Set(ref, newRecord(replaceWith, b.clock)); err != nil {
-			return ConvertGRPCError(err)
+	docSnap, err := b.svc.Collection(b.CollectionName).Doc(b.keyToDocumentID(expected.Key)).Get(ctx)
+	if err != nil {
+		err = ConvertGRPCError(err)
+		if trace.IsNotFound(err) {
+			return nil, trace.CompareFailed(err.Error())
 		}
+		return nil, trace.Wrap(err)
+	}
 
-		return nil
-	})
+	r, err := newRecordFromDoc(docSnap)
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
 
-	return b.newLease(replaceWith), nil
+	if !bytes.Equal(r.Value, expected.Value) {
+		return nil, trace.CompareFailed("expected item value %v does not match actual item value %v", string(expected.Value), r.Value)
+	}
+
+	replaceWith.Revision = expected.Revision
+	return b.ConditionalUpdate(ctx, replaceWith)
```

Comment:
I'm a bit worried that this might lead to spurious `CompareAndSwap` failures that are not part of the backend contract, especially for `CompareAndSwapCertAuthority` as used by `RotateCertAuthority` and `RotateExternalCertAuthority`.

Can we give it a small handful of retries?

Reply:
A CompareFailed error returned from CompareAndSwap, in the current contract, means "the value does not exist or the value exists and is different from the expected value". A keepalive or any write done "just in case" with the same value as the existing value will cause a concurrent CompareAndSwap to fail even if it was supposed to succeed.

I don't feel too strongly about this, especially because we can remove CompareAndSwap quite easily (and we should) once ConditionalUpdate is in place, but for the same reason we shouldn't bother with changing the contract and figuring out if the change to "the value does not exist or the value exists and is different from the expected value or the backend had a spurious failure and the operation can be retried" will cause issues in existing code.